kale/embed/_mpca_torch.py

```diff
@@ -0,0 +1,230 @@
+# =============================================================================
+# Author: Shuo Zhou, szhou20@sheffield.ac.uk
+#         Haiping Lu, h.lu@sheffield.ac.uk or hplu@ieee.org
+# =============================================================================
+
+"""Multilinear Principal Component Analysis (MPCA) implementation in Pytorch
+
+Reference:
+    Haiping Lu, K.N. Plataniotis, and A.N. Venetsanopoulos, "MPCA: Multilinear
+    Principal Component Analysis of Tensor Objects", IEEE Transactions on Neural
+    Networks, Vol. 19, No. 1, Page: 18-39, January 2008. For initial Matlab
+    implementation, please go to https://uk.mathworks.com/matlabcentral/fileexchange/26168.
+"""
+# import pytorch_lightning as pl
+import warnings
+
+import tensorly as tl
+import torch
+import torch.nn as nn
+
+# from tensorly import fold, unfold
+from tensorly.tenalg import multi_mode_dot
+
+from ._mpca import _check_dim_shape
+
+tl.set_backend("pytorch")
+
+
+class MPCA(nn.Module):
+    """MPCA implementation compatible with pytorch
+
+    Args:
+        var_ratio (float, optional): Percentage of variance explained
+            (between 0 and 1). Defaults to 0.97.
+        max_iter (int, optional): Maximum number of iteration. Defaults to 1.
+
+
+    Attributes:
+        factors (list of array-like): A list of transposed projection matrices, shapes (P_1, I_1), ...,
+            (P_N, I_N), where P_1, ..., P_N are output tensor shape for each sample.
+        idx_order (tensor): The ordering index of projected (and vectorised) features in decreasing variance.
+        # mean_ (tensor): Per-feature empirical mean, estimated from the training set, shape (I_1, I_2, ..., I_N).
+        shape_in (tuple): Input tensor shapes, i.e. (I_1, I_2, ..., I_N).
+        shape_out (tuple): Output tensor shapes, i.e. (P_1, P_2, ..., P_N).
+    """
+
+    def __init__(self, var_ratio=0.97, max_iter=1):
+        super().__init__()
+
+        self.var_ratio = var_ratio
+        if max_iter > 0 and isinstance(max_iter, int):
+            self.max_iter = max_iter
+        else:
+            raise ValueError(
+                "Number of max iterations must be an integer and greater " "than 0, but given %s" % max_iter
+            )
+        self.factors = []
+        self.shape_in = ()
+        self.shape_out = ()
+        # self.mean_ = None
+        self.idx_order = None
+        self.n_dim = None
+
+    def _init_model(self, x):
+        """Init model parameters for x"""
+        shape_ = x.shape  # shape of input data
+        n_spl = shape_[0]  # number of samples
+        n_dim = x.ndim
+
+        # self.mean_ = torch.mean(x, 0)
+        # x_ = x - self.mean_
+
```

Comment:
Why leaving mean out here but not in (_)mpca.py?

Reply:
Because pytorch Dataloader seems normalise data in a different way, the mean of training examples of each batch have been removed when they are passed to the training model. I am not sure about this, so I just made these code comments.

Reply:
Add part of your reply to the docstrings may help since mean removal is an important step in (M)PCA.
Here it should be OK if the mean is already zero. That's why it will be still good for us to check how consistent the two versions of MPCA are.